src/passes/Intrinsics.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/intrinsics.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+struct IntrinsicLowering : public WalkerPass<PostWalker<IntrinsicLowering>> {
+  bool isFunctionParallel() override { return true; }
+
+  Pass* create() override { return new IntrinsicLowering; }
+
+  void visitCall(Call* curr) {
+    if (Intrinsics(*getModule()).isCallWithoutEffects(curr)) {
+      // Turn into a call, by using the final operand as the function to call.
+      auto& operands = curr->operands;
+      auto* target = operands.back();
+      operands.pop_back();
+      // We could rely on later optimizations here, but at least ensure we emit
+      // a direct call when we can, to avoid a performance cliff if the user
+      // forgets to optimize.
+      Builder builder(*getModule());
+      if (auto* refFunc = target->dynCast<RefFunc>()) {
+        auto name = refFunc->func;
+        auto* func = getModule()->getFunction(name);
+        if (!HeapType::isSubType(func->type, refFunc->type.getHeapType())) {
+          Fatal() << "call.without.effects is not of a subtype of the called "
+                     "function ("
+                  << name << ")\n";
+        }
```

Comment:
Isn't this a part of the validation of `RefFunc`, separate from `call.without.effects`, which should have been done in the validator already?

Reply:
Good catch! I wanted to check that the type of the `call.without.effects` is the right one. It should be identical to the function reference's, ignoring the last parameter (which is the function reference), but this code does not do that...

Actually checking this would be a bunch more code, so I think maybe I'll just remove it, and leave it to the validator. I'd hoped for a clearer error message here, but let's see if it's a problem in practice.